wallet-client/Main.hs

```diff
@@ -0,0 +1,126 @@
+                                      CardanoEnvironment, CardanoFeature (..),
+                                      CardanoFeatureInit (..),
+                                      initializeCardanoEnvironment)
+
+import           CLI
+import           Run
+
+-- | The product type of all command line arguments
+data ArgParser = ArgParser !CLIarguments !CLI
+
+-- | The product parser for all the CLI arguments.
+--
+commandLineParser :: Parser ArgParser
+commandLineParser = ArgParser
+    <$> parse
+    <*> parseCLI
+
+-- | Top level parser with info.
+--
+opts :: ParserInfo ArgParser
+opts = info (commandLineParser <**> helper)
+  ( fullDesc
+  <> progDesc "Cardano wallet node."
+  <> header "Demo client to run.")
+
+-- | Main function.
+main :: IO ()
+main = do
+
+    let cardanoConfiguration = mainnetConfiguration
+    cardanoEnvironment  <- initializeCardanoEnvironment
+
+    logConfig           <- execParser opts
+
+    (cardanoFeatures, nodeLayer) <- initializeAllFeatures logConfig cardanoConfiguration cardanoEnvironment
+
+    let cardanoApplication :: NodeLayer -> CardanoApplication
+        cardanoApplication = CardanoApplication . nlRunNode
+
+    runCardanoApplicationWithFeatures Development cardanoFeatures (cardanoApplication nodeLayer)
+
+initializeAllFeatures :: ArgParser -> CardanoConfiguration -> CardanoEnvironment -> IO ([CardanoFeature], NodeLayer)
+initializeAllFeatures (ArgParser logCli cli) cardanoConfiguration cardanoEnvironment = do
+
+    (loggingLayer, loggingFeature) <- createLoggingFeature cardanoEnvironment cardanoConfiguration logCli
+    (nodeLayer   , nodeFeature)    <- createNodeFeature loggingLayer cli cardanoEnvironment cardanoConfiguration
+
+    -- Here we return all the features.
+    let allCardanoFeatures :: [CardanoFeature]
+        allCardanoFeatures =
+            [ loggingFeature
+            , nodeFeature
+            ]
+
+    pure (allCardanoFeatures, nodeLayer)
+
+--------------------------------
+-- Layer
+--------------------------------
+
+data NodeLayer = NodeLayer
+    { nlRunNode   :: forall m. MonadIO m => m ()
+    }
+
+--------------------------------
+-- Node Feature
+--------------------------------
+
+-- type NodeCardanoFeature = CardanoFeatureInit LoggingLayer Text NodeLayer
+type NodeCardanoFeature = CardanoFeatureInit LoggingLayer CLI NodeLayer
+
+
+createNodeFeature :: LoggingLayer -> CLI -> CardanoEnvironment -> CardanoConfiguration -> IO (NodeLayer, CardanoFeature)
+createNodeFeature loggingLayer cli cardanoEnvironment cardanoConfiguration = do
+    -- we parse any additional configuration if there is any
+    -- We don't know where the user wants to fetch the additional configuration from, it could be from
+    -- the filesystem, so we give him the most flexible/powerful context, @IO@.
+
+    -- we construct the layer
+    nodeLayer <- (featureInit nodeCardanoFeatureInit) cardanoEnvironment loggingLayer cardanoConfiguration cli
```

Comment:
What's with the parens? Isn't it the same without them?

Reply:
+1, I'll remove them. The sin of copy pasting :D